usr/share/rear/lib/_input-output-functions.sh

```diff
@@ -1177,6 +1177,28 @@ function UserInput () {
     # In debug mode show the user the script that called UserInput and what user_input_ID was specified
     # so that the user can prepare an automated response for that UserInput call (without digging in the code):
     DebugPrint "UserInput -I $user_input_ID needed in $caller_source"
+    # Drain stdin if stdin is a terminal i.e. when 'rear' is run in interactive mode
+    # where stdin is all what the user types on his keyboard.
+    # In this case discard possibly already existing characters (in particular ENTER keystrokes) from stdin
+    # to avoid that when the user had accidentally hit ENTER more than once in a previous (UserInput) dialog
+    # then those additional ENTER characters would be already in stdin and let this current UserInput dialog
+    # proceed unintendedly automatically without an explicit ENTER from the user for this current dialog,
+    # see https://github.com/rear/rear/issues/2866
+    # There is no generic way to clear stdin so we 'read' and discard what is already there
+    # up to 1000 characters (so it fails if there are more than 1000 characters in stdin)
+    # see https://superuser.com/questions/276531/clear-stdin-before-reading
+    # That the 'read' timeout can be a fractional number requires bash 4.x
+    # but in general ReaR should still work with bash 3.x so we use '-t 1'
+    # which causes a one second delay (in interactive mode) which cannot be avoided,
+    # see https://github.com/rear/rear/issues/2866#issuecomment-1254908270
+    local discard_stdin=""
+    if tty -s ; then
+        # This 'read' call usually exits with non-zero exit code
+        # (unless it gets valid input terminated by ENTER):
+        read -s -t 1 -n 1000 discard_stdin
```

Comment:
I am afraid that this drains only the first ENTER from the input and that you should use `-N` instead of `-n`.

Reply:
Thank you for checking it.
You are right:
On openSUSE Leap 15.3 with 4.4.23
```
# echo -e 'One\nTwo\nThree' | ( read -t 1 -n 10 D ; read R1 ; read R2 ; echo D "$D" ; echo R1 "$R1" ; echo R2 "$R2" )
D One
R1 Two
R2 Three

# echo -e 'One\nTwo\nThree' | ( read -t 1 -N 10 D ; read R1 ; read R2 ; echo D "$D" ; echo R1 "$R1" ; echo R2 "$R2" )
D One
Two
Th
R1 ree
R2 
```

But SLES11 SP4 with bash version 3.2.57
'read' does not support `-N`:
```
# echo -e 'One\nTwo\nThree' | ( read -t 1 -N 10 D ; read R1 ; read R2 ; echo D "$D" ; echo R1 "$R1" ; echo R2 "$R2" )
-bash: read: -N: invalid option
read: usage: read [-ers] [-u fd] [-t timeout] [-p prompt] [-a array] [-n nchars] [-d delim] [name ...]
D 
R1 One
R2 Two
```

What works for me even with bash 3.x is
using an artificial delimiter character
that is sufficiently unlikely
to be accidentally entered by the user:
```
# echo -e 'One\nTwo\nThree' | ( read -t 1 -d '~' -n 10 D ; read R1 ; read R2 ; echo D "$D" ; echo R1 "$R1" ; echo R2 "$R2" )
D One
Two
Th
R1 ree
R2 
```

Reply:
what could then work is : `read -s -t 0 -N 1000 discard_stdin || read -s -t 1 -n 1000 discard_stdin` (read with `-t 1` and `-n` would serve as a fallback for the bash 3 case, of course in bash).

Reply:
Or, better and simpler: use a loop:  `while read -s -t 1 -n 1000 discard_stdin; done`

Reply:
test:
```$ echo -e 'One\nTwo\nThree' | ( while read -s -t 1 -n 1000 D; do  echo D "$D" ; done;  read R1 ; read R2 ; echo R1 "$R1" ; echo R2 "$R2" )
D One
D Two
D Three
R1 
R2 
```

Reply:
not tested with bash 3 though

Reply:
On SLES11 SP4 with bash version 3.2.57
```
# time echo -e 'One\nTwo\nThree' | ( while read -s -t 1 -n 1000 D ; do echo D "$D" ; done ; read R1 ; read R2 ; echo R1 "$R1" ; echo R2 "$R2" )
D One
D Two
D Three
R1 
R2 

real    0m0.002s
user    0m0.000s
sys     0m0.000s

# time echo -en '' | ( while read -s -t 1 -n 1000 D ; do echo D "$D" ; done ; read R1 ; read R2 ; echo R1 "$R1" ; echo R2 "$R2" )
R1 
R2 

real    0m0.002s
user    0m0.004s
sys     0m0.000s

# time while read -s -t 1 -n 1000 D ; do echo D "$D" ; done

real    0m1.000s
user    0m0.000s
sys     0m0.000s
```
same timing behaviour on openSUSE Leap 15.3 with bash 4.4.23

I had expected an endless loop of
```
while read -s -t 1 -n 1000 D ; do echo D "$D" ; done
```
but fortunately 'read' exits with non-zero exit code
when read times out which breaks the while loop.

Reply:
I think the `-n 1000` is not needed.
I wonder why it was used at all?

Meanwhile it seems
```
test -t 0 && read -t 0 && while read -s -t 1 ; do : ; done
```
is all in one line and avoids the one second delay.

Reply:
`read -t 0` will work, but not with bash 3. Without `-n 1000`, this won't drain input not terminated by ENTER (the user may have hit another character, even if it is less likely and less dangerous).

Reply:
(well I have not tested that `-t 0` will not work with bash 3, that's what I have read )